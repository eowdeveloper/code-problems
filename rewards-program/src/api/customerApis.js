const apiUrl = 'http://localhost:3000';

export const fetchTransactions = async () => {
  try {
    const response = await fetch(`${apiUrl}/transactions`);
    const data = await response.json();
    return data.transactions;
  } catch (error) {
    console.error('Error fetching transactions:', error);
  }
};
