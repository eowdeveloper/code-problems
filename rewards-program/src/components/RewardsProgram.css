body {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 20px;
}

h1 {
  font-size: 30px;
  text-align: center;
}

table {
  border-collapse: collapse;
  margin-top: 20px;
  width: max-content;
}

th, td {
  padding: 20px;
  text-align: center;
  border: 1px solid #ddd;
}

th {
  background-color: #f2f2f2;
  font-weight: bold;
}

tbody tr:hover {
  background-color: #f9f9f9;
  cursor: pointer;
}

.loading {
  font-size: 24px;
  text-align: center;
  margin-top: 50px;
}
