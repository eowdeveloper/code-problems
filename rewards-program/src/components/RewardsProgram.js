import { useState, useEffect } from "react";
import { fetchTransactions } from "../api/customerApis";
import "./RewardsProgram.css";

function RewardsProgram() {
  const [transactions, setTransactions] = useState([]);
  const [rewards, setRewards] = useState([]);

  useEffect(() => {
    fetchTransactions().then((data) => {
      setTransactions(data);
    }).catch((error) => {
      console.error("Error fetching transactions:", error);
      setTransactions([]);
    });
  }, []);

  useEffect(() => {
    const calculateRewards = (transactions) => {
      const rewards = [];

      transactions.forEach((transaction) => {
        const { customer, date, amount } = transaction;
        const month = date.slice(0, 7);
        let rewardPoints = 0;

        if (amount > 100) {
          rewardPoints += (amount - 100) * 2 + 100;
        } else if (amount > 50) {
          rewardPoints += amount - 50;
        }

        const existingReward = rewards.find(
          (reward) => reward.customer === customer && reward.month === month
        );

        if (existingReward) {
          existingReward.points += rewardPoints;
        } else {
          rewards.push({ customer, month, points: rewardPoints });
        }
      });

      setRewards(rewards);
    };

    calculateRewards(transactions);
  }, [transactions]);

  const renderResults = () => {
    return (
      <table>
        <thead>
          <tr>
            <th>Customer</th>
            <th>Month</th>
            <th>Reward Points</th>
          </tr>
        </thead>
        <tbody>
          {rewards.map((reward, index) => {
            return (
              <tr key={index}>
                <td>{reward.customer}</td>
                <td>{reward.month}</td>
                <td>{reward.points}</td>
              </tr>
            );
          })}
        </tbody>
      </table>
    );
  };

  return (
    <div>
      <h1>Reward Points Calculator</h1>
      {transactions.length === 0 ? (
        <p className="loading">Loading...</p>
      ) : (
        renderResults()
      )}
    </div>
  );
}

export default RewardsProgram;