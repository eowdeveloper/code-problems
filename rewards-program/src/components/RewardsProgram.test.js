import { render, screen } from "@testing-library/react";
import RewardsProgram, {calculateRewards, sortingMonth} from "./RewardsProgram";
import * as api from "../api/customerApis";

describe("RewardsProgram", () => {
  const transactions = [
    { id: 1, name: "Adam", amount: 120, date: 1673249112 },
    { id: 2, name: "Adam", amount: 50, date: 1675668312 },
    { id: 3, name: "Joe", amount: 80, date: 1674199512 },
    { id: 4, name: "Adam", amount: 60, date: 1678688712 },
    { id: 5, name: "Joe", amount: 130, date: 1676273112 },
    { id: 6, name: "Joe", amount: 70, date: 1679984712 },
  ];

  beforeEach(() => {
    jest.spyOn(api, "fetchTransactions").mockResolvedValue(transactions);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test("renders table with reward points after transactions are fetched", async () => {
    render(<RewardsProgram />);
    await screen.findByText("Reward Points");

    const table = screen.getByRole("table");

    expect(table).toMatchSnapshot();
  });

  test("calculateRewards returns correct rewards for transactions", () => {
    const rewards = calculateRewards(transactions);
    expect(rewards).toMatchSnapshot();
  });

  test("sortingMonth returns correct months array for transactions", () => {
    const monthArr = sortingMonth(transactions);
    expect(monthArr).toMatchSnapshot();
  });
});

