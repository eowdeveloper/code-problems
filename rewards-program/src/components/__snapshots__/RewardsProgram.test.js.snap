// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RewardsProgram renders table with reward points after transactions are fetched 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Total Points
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
